config/v1/types_feature.go

```diff
@@ -120,12 +120,14 @@ var FeatureSets = map[FeatureSet]*FeatureGateEnabledDisabled{
 
 var defaultFeatures = &FeatureGateEnabledDisabled{
 	Enabled: []string{
-		"APIPriorityAndFairness",         // sig-apimachinery, deads2k
-		"RotateKubeletServerCertificate", // sig-pod, sjenning
-		"SupportPodPidsLimit",            // sig-pod, sjenning
-		"NodeDisruptionExclusion",        // sig-scheduling, ccoleman
-		"ServiceNodeExclusion",           // sig-scheduling, ccoleman
-		"SCTPSupport",                    // sig-network, ccallend
+		"APIPriorityAndFairness",            // sig-apimachinery, deads2k
+		"RotateKubeletServerCertificate",    // sig-pod, sjenning
+		"SupportPodPidsLimit",               // sig-pod, sjenning
+		"NodeDisruptionExclusion",           // sig-scheduling, ccoleman
+		"ServiceNodeExclusion",              // sig-scheduling, ccoleman
+		"SCTPSupport",                       // sig-network, ccallend
+		"EndpointSliceProxying",             // sig-network, mmasters
```

Comment:
@Miciah since you are updating all the EndpointSlices feature sets may I ask you to enable the `EndpointSliceNodeName` feature too?

https://github.com/kubernetes/kubernetes/blob/5b8c3b90f3c74ba79f4bfa5be560d011694a7da0/pkg/features/kube_features.go#L804

you can add me or @danwinship , he already has the IPv6DualStack so maybe better him

xref: https://github.com/kubernetes/kubernetes/pull/96440

Reply:
Last I asked @danwinship, he said we weren't going to enable `EndpointSliceProxying` "until upstream Kube converts".  However, it looks like upstream has enabled `EndpointSliceProxying` by default in 1.19:  <https://github.com/kubernetes/kubernetes/pull/92736/commits/8039cf9bb1a3d3c600c0eaf35b53b008f0226ffc>.  

Moreover, as far as I can tell, enabling these feature gates in openshift/api does not have any effect; what would have an effect would be enabling these feature gates in <https://github.com/openshift/sdn/blob/master/vendor/k8s.io/kubernetes/pkg/features/kube_features.go>.  This means that `EndpointSliceProxying` will be enabled once openshift/sdn is rebased to Kubernetes 1.20, unless someone explicitly patches it off; and we'll need to patch openshift/sdn to enable `EndpointSliceNodeName` if and when we agree to enable that feature in OpenShift.

Reply:
got it, thanks for the clarification

Reply:
> Moreover, as far as I can tell, enabling these feature gates in openshift/api does not have any effect;

Yes, openshift-sdn does not currently track the `FeatureGate` object, and so ignores the definitions here (SDN-1325). But that doesn't matter to _you_ because you don't care whether sdn/kube-proxy supports EndpointSlices, you only care whether kube-apiserver and kube-controller-manager support EndpointSlice. So enabling the feature gates here is the correct approach, because the operators for apiserver and kcm do track the definitions here. (Well, you'll need to make sure `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` revendor `openshift/api` after this merges.)

That said, `EndpointSliceTerminatingCondition` and `EndpointSliceNodeName` are both alpha so I'm not sure what our rules are about enabling them.

Reply:
> Yes, openshift-sdn does not currently track the `FeatureGate` object, and so ignores the definitions here ([SDN-1325](https://issues.redhat.com/browse/SDN-1325)). But that doesn't matter to _you_ because you don't care whether sdn/kube-proxy supports EndpointSlices, you only care whether kube-apiserver and kube-controller-manager support EndpointSlice.

I need `EndpointSliceProxying` for [https://issues.redhat.com/browse/NE-348](NE-348) (graceful termination for load balancers with "Local" external traffic policy).  

You might be thinking of [https://issues.redhat.com/browse/NE-376](NE-376) (dual-stack support in the router), which requires `EndpointSlice`, but `EndpointSlice` is already enabled by default.

Reply:
Oh, I was confused about names. I thought `EndpointSliceProxying` was the gate for `endpointslicemirroring.Controller`... Didn't realize there was a separate gate for "kube-proxy uses EndpointSlices".

Anyway, openshift-sdn does not support EndpointSlice, and there's no way it can be made to use it currently. However, it looks to me like in 1.20, the upstream kube-proxy code does not check the `EndpointSliceTerminatingCondition` feature gate, or ever look at the `Terminating` condition in `EndpointSlice`, so it seems like that KEP isn't fully implemented in 1.20 anyway? (It looks like the API is all there, but nothing uses it yet.)

(FTR, I filed SDN-1382 about porting openshift-sdn to EndpointSlice in 4.8 or 4.9, because I realized upstream is going to want to remove the legacy Endpoints code from kube-proxy once EndpointSlice goes GA anyway, so we're going to have to deal with it.)

Reply:
> Anyway, openshift-sdn does not support EndpointSlice, and there's no way it can be made to use it currently. However, it looks to me like in 1.20, the upstream kube-proxy code does not check the `EndpointSliceTerminatingCondition` feature gate, or ever look at the `Terminating` condition in `EndpointSlice`, so it seems like that KEP isn't fully implemented in 1.20 anyway? (It looks like the API is all there, but nothing uses it yet.)

Right, the change to the `EndpointSlice` API and controller is in 1.20, but the change to kube-proxy ([upstream 96371](https://github.com/kubernetes/kubernetes/pull/96371)) didn't make it in; it looks like it will be in 1.21.  

> (FTR, I filed [SDN-1382](https://issues.redhat.com/browse/SDN-1382) about porting openshift-sdn to EndpointSlice in 4.8 or 4.9, because I realized upstream is going to want to remove the legacy Endpoints code from kube-proxy once EndpointSlice goes GA anyway, so we're going to have to deal with it.)

Excellent!  I've made [NE-478](https://issues.redhat.com/browse/NE-478) to track the remaining work for graceful termination and linked it to SDN-1382.